### PR TITLE
Bidirectional version of Dijkstra's shortest path

### PR DESCRIPTION
**Pull request by d-michail**
This is an implementation of a bidirectional version of Dijkstra's shortest path algorithm (closes issue #135).

The algorithm maintains two frontiers and alternates between them until they meet.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+    public GraphPath<V, E> run()
+    {
+        // handle special case if source equals target
+        if (source.equals(target)) {
+            return new GraphWalk<>(
+                forwardFrontier.graph,
+                source,
+                target,
+                Arrays.asList(source),
+                new ArrayList<>(),
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java]: Consider: Collections.emptyList()

File: jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+
+    public GraphPath<V, E> run()
+    {
+        // handle special case if source equals target
+        if (source.equals(target)) {
+            return new GraphWalk<>(
+                forwardFrontier.graph,
+                source,
+                target,
+                Arrays.asList(source),
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java]: Consider Collections.singletonList(source)

File: jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+    private final V target;
+    private final double radius;
+
+    public BidirectionalDijkstraShortestPathDetails(
+        Graph<V, E> graph,
+        V source,
+        V target,
+        double radius)
+    {
+        this.forwardFrontier = new SearchFrontier<V, E>(graph);
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java]: Consider replacing <V,E> by <> (also elsewhere, no longer needed in java 8)

File: jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+     *
+     * @param g the graph for which this specifics object to be created.
+     */
+    public UndirectedSpecifics(Graph<VV, EE> g)
+    {
+        graph = g;
+    }
+
+    /**
+     * @see CrossComponentIterator.Specifics#edgesOf(Object)
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java]: Not sure why you reference to the CrossComponentIterator here? I think that the qualifier "import org.jgrapht.traverse.CrossComponentIterator;" is also missing.

File: jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+
+/**
+ * Private data to associate with each entry in the priority queue.
+ */
+class QueueEntry<V, E>
+{
+    E e;
+    V v;
+
+    QueueEntry()
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java]: empty constructor may be omitted.

File: jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+
+}
+
+/**
+ * The implementation details
+ */
+class BidirectionalDijkstraShortestPathDetails<V, E>
+{
+    private final SearchFrontier<V, E> forwardFrontier;
+    private final SearchFrontier<V, E> reversedFrontier;
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java]: Consider backwardFrontier (as opposite of forwardFrontier)

**Comment by jkinable** [comment]
Looks good to me. Several people have asked for a Bidirectional shortest path in the past, so this is a nice addition. I've included some minor remarks. Let's wait to see whether @jsichi has any additional remarks.

**Comment by d-michail** [comment]
Thanks for the comments. I will fix them as soon as possible. 

Regarding the class structure, the public class BidirectionalDijkstraShortestPath has the same structure as DijkstraShortestPath. If you see the test cases, I have an example where I execute both and test whether the result is the same.

Class BidirectionalDijkstraShortestPathDetails is not exported outside the file. I wrote it like that in order to allow the garbage collector to regain used memory even if the caller keeps a BidirectionalDijkstraShortestPath instance around for a long time.

**Comment by d-michail** [comment]
Fixed the minor issues. 

I also added a static findPathBetween(V a, V b) method in the BidirectionalDijkstraShortestPath class. Now the interface is exactly the same as in DijkstraShortestPath.

**Comment by jkinable** [comment]
Looks good! Btw, nice job on the extensive unit testing and proper documentation.

In the future it would be nice if findPathBetween in both Dijkstra classes, would actually return a GraphPath as the name suggests (instead of a vertex path). That way the user has instant access to a full description of the path in terms of edges, vertices and cost. That would also work for multigraphs. For now I suggest we keep it like it is.

**Comment by jsichi** [comment]
lgtm.  @jkinable do you want to go ahead and merge the pull request?

(As a followup, perhaps we should add a new interface to org.jgrapht.alg.interfaces to make the commonality explicit?)

Regarding the findPathBetween convenience method, we could deprecate it and add the improved version to java.org.jgrapht.Graphs (with an extra parameter controlling whether to use bidirectional search).

**Comment by jkinable** [comment]
Thanks @d-michail , nice contribution!

We should indeed add a new interface to org.jgrapht.alg.interfaces which covers single source-target paths.

**Pull request merged**